### PR TITLE
Fast & lightweight tempcontrol

### PR DESCRIPTION
**Pull request by LuKePicci**
Before: at least two nvidia-smi calls per gpu, plus one nvidia-settings call per gpu
On my rig: 22 nvidia-smi calls per round, plus 11 nvidia-settings calls per round

After: usually one nvidia-smi call, plus one nvidia-settings call
On my rig: 1 nvidia-smi call per round, plus 1 nvidia-settings call per round

There was also a bug that showed up with my custom tempcontrol settings which allowed fans to be set to something below minimum while powerlimit continued to be decreased because of high temperature.

### CODE REVIEW

**Comment by papampi** [comment]
Have you tested it?
On my test rig it opens nvidia-settings window on every round and not continue unless I close it, then at next round it opens it again.
```
GPU 0:  POWER LIMIT: 100,  TARGET TEMP: 70
GPU 1:  POWER LIMIT: 75,  TARGET TEMP: 70
GPU 2:  POWER LIMIT: 75,  TARGET TEMP: 70

GLOBAL  FAN_ADJUST (%):  2
GLOBAL  POWER_ADJUST (W):  2
GLOBAL  ALLOWED_TEMP_DIFF (C):  2
GLOBAL  RESTORE_POWER_LIMIT (%):  80
GLOBAL  MINIMAL_FAN_SPEED (%):  50

GPU 0, Target temp: 70, Current: 59, Diff: 11, Fan: 50, Power: 11.26
GPU 1, Target temp: 70, Current: 55, Diff: 15, Fan: 50, Power: 6.59
GPU 2, Target temp: 70, Current: 62, Diff: 8, Fan: 50, Power: 7.17
```

After last line nvidia-settings opens and stay open, when I close it these lines show up and nvidia-settings open again:

```
Fri Jul 27 16:50:06 +0430 2018 - All good, will check again in 18.5 seconds

GPU 0, Target temp: 70, Current: 59, Diff: 11, Fan: 50, Power: 100.28
GPU 1, Target temp: 70, Current: 58, Diff: 12, Fan: 50, Power: 73.18
GPU 2, Target temp: 70, Current: 64, Diff: 6, Fan: 50, Power: 75.41
```

**Comment by LuKePicci** [comment]
My rig is running fine since yesterday on the pluggable miners branch with no problems. I will now test this one and see what happens. Try editing the "sudo $NVD} …" line at the end such that it also echo the complete command line before calling nvidia-settings, let's see why it stays open.

**Comment by LuKePicci** [comment]
Got it, my bad, I forgot to clean the variable at each cycle, and never got stuck because it never happened to me that on first round no fan had to be adjusted.

**Comment by papampi** [comment]
Nvidia-settings problem is solved.
But now its allowing temp to go alowed_diff degree higher than target and not fixing it on target_temp.
It should higher fan when temp goes above target_temp, even 1 degree.
My allowed diff is 2:

```
GPU 0, Target temp: 60, Current: 62, Diff: 2, Fan: 54, Power: 92.17
GPU 1, Target temp: 60, Current: 62, Diff: 2, Fan: 46, Power: 74.79
GPU 2, Target temp: 60, Current: 61, Diff: 1, Fan: 70, Power: 74.54
Fri Jul 27 18:23:56 +0430 2018 - All good, will check again in 18.5 seconds

```


Edit:
In 1bash we have it too
`ALLOWED_TEMP_DIFF=2           # If current temp is 2C below the target temp reduce the fan speed. Works only if current temp is below target temp
`

File: 6tempcontrol [hunk 1]
```diff
@@ -297,22 +301,27 @@ do
...
     POWERLIMIT=${PWRLIMIT%%.*}
-    TEMP_DIFF=$(( TARGET_TEMP[GPU] - CURRENT_TEMP ))
+    TEMP_DIFF=$(( CURRENT_TEMP - TARGET_TEMP[GPU] ))
     NEW_FAN_SPEED=$CURRENT_FAN
     HIGH_TEMP_DIFF=$(( ALLOWED_TEMP_DIFF * 2 ))
 
     echo -e "${B}GPU $GPU${N}, Target temp: ${B}${TARGET_TEMP[${GPU}]}${N}, Current: ${B}$CURRENT_TEMP${N}, Diff: ${B}$TEMP_DIFF${N}, Fan: ${B}$CURRENT_FAN${N}, Power: ${B}$POWERDRAW${N}"
 
-    if (( CURRENT_TEMP > TARGET_TEMP[GPU] ))
+    if (( TEMP_DIFF > ALLOWED_TEMP_DIFF ))
```
Comment (papampi) [inline 6tempcontrol]: This should be reverted back to the way it was, so that we can fix on target and not letting it go above.

**Comment by LuKePicci** [comment]
Now should match the behavior described in 1bash. I've got confused because ALLOWED_TEMP_DIFF is also used to generate HIGH_TEMP_DIFF which is used as an upward margin to accelerate fans speed up.

**Review by papampi — APPROVED** [review]

**Comment by papampi** [comment]
We can use same GPU_NAME in watchdog too and reduce nvidia-smi calls

**Comment by papampi** [comment]
What do you think about nvidia persistance mode?
```# Setting persistance mode on, will keep settings in between sessions
 # fullzero, Papampi, Stubo..., we need to decide if we gonna use Persistane Mode.	 
 # I am using it for a while and had no problems with it.	 
 # If we decide to use it, please move this to the top of 3main:	 
 # sudo nvidia-smi -pm 1	 ```

From ubuntu man pages:
       The  nvidia-persistenced utility is used to enable persistent software state in the NVIDIA
       driver.  When persistence mode is enabled, the daemon prevents the driver  from  releasing
       device  state  when  the  device  is not in use.  This can improve the startup time of new
       clients in this scenario.

**Comment by papampi** [comment]
Just tested persistence mode on my test rig and it feels like tempcontrol loads much faster, we can either enable it by default or add a 1bash entry for it.

**Comment by LuKePicci** [comment]
The persistence functionality is distro-specific.  Since nvidia-persistenced is a systemd service has to be enabled once for all, we don't need to enable it each time at the beginning of 3main. In our Ubuntu nvOC pre-built images we have Xorg as a client for each gpu, so their driver instance does not get unloaded. It does when all existing clients close the gpu device file, but when we close the miner, Xorg is still there so the driver never gets unloaded. That's why OC settings survive across miner launch/exit. otherwise simply restarting the miner would have cleared OC settings to their defaults. If we build an image where xorg is not a client for each gpu then we will have to enable nvidia-persisteced daemon.

**Comment by LuKePicci** [comment]
@papampi you can't observe any faster loading by enabling nvidia-persistenced to keep the gpu device file open because Xorg and compiz are keeping this file open already. 

```
m1@m1-desktop:/dev$ sudo lsof nvidia0
COMMAND     PID                USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
Xorg        864                root  mem    CHR  195,0           449 nvidia0
Xorg        864                root   12u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   13u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   14u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   58u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   59u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   61u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   62u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root   63u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root  126u   CHR  195,0      0t0  449 nvidia0
Xorg        864                root  127u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1  mem    CHR  195,0           449 nvidia0
compiz     2193                  m1   12u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1   13u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1   37u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1   38u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1   39u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1   40u   CHR  195,0      0t0  449 nvidia0
compiz     2193                  m1   41u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  mem    CHR  195,0           449 nvidia0
zm_miner  29112                  m1    5u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1   16u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1   17u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1   41u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1   52u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1   53u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  115u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  123u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  135u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  169u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  172u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  174u   CHR  195,0      0t0  449 nvidia0
zm_miner  29112                  m1  176u   CHR  195,0      0t0  449 nvidia0
```
If you enable nvidia-persisteced one more line appears:
```
nvidia-pe  1067 nvidia-persistenced    5u   CHR  195,0      0t0  449 nvidia0
```
If I stop zm_miner there are still Xorg, compiz and nvidia-persisteced. If we detache Xorg an compiz from al gpus except gpu0 then nvidia-persistenced will be the only remaining file handle and driver will remain loaded because of this one only.

**Comment by leenoox** [comment]
If I remember right xorg is not a client for nvidia. An app must use nvidia
to keep the P-level from resetting, xorg alone won't keep the settings. To
achieve this I think we added endless loop at the end of 3main. Remove that
endless loop and P-level will reset creating all sorts of problems,
hanging, freezing since the applied OC values for P2  will be applied to
different P level which will be way beyond what GPU can handle and it will
freeze. Enabling persistence mode at startup will resolve that problem and
the endless loop won't be needed.
I think it is best to stick with distro specific nvoc to prevent all sorts
of problems... although it is nice to have nvoc on any distro it will
create too much problems, different versions of software, kernel, settings,
etc.

**Comment by LuKePicci** [comment]
Xorg is absolutely a client of type 'G' (ie graphic, I think…), and it is listed in nvidia-smi list of clients as well. nVidia documentation is clear about that, the persistence functionality born to avoid driver unloading on HPC multi-gpu environments when all clients of type type 'C' (ie. cuda, I guess) stop and there is no Xorg running.
The same documentation describes quite well how the nvidia-persistenced is implemented (it is open source under MIT license, if you would like to see how it works). It only opens the device file and sleeps. I don't know what kind of problems you had with performace states. Maybe some gpus prevent certain P states when you run unsupported job types, like graphic jobs on nvidia Tesla gpus?

**Comment by leenoox** [comment]
I guess C is for compute. Miners force the driver in compute mode and it
locks at P2. When the switcher decides to mine different coin it kills the
miner to launch another one. I think it works like this... When the miner
is killed, driver exits the compute mode and switches and cycles through P
levels, briefly going to P0 then P8 before the other miner is launched. The
applied OC values remain while only P level is changed. While P2 can handle
+1000 memory OC, P8 can handle only +600 and freezes, it is same as you
would apply +1400 at P2 while GPU can handle only +1000. Dont quote me on
this, I might be wrong about the P levels but in essence it is how it
works. We added endless loop to keep the GPU's locked at P2, parrent-child
processes.
Sorry if I'm wrong and mislead you, it's been a while since I looked at
persistence mode but I believe it worked as described above...

**Comment by LuKePicci** [comment]
I think you're right about the P level switching, but OC values we apply are specific to certain P profiles, I mean, only P2, for example, have customizable OC values, so the actual frequency change from levels set for P2 to levels set for P0 when the compute job stops. This is normal, they come back as soon as a new compute job starts and P switches again to P2. I'm sure all gpus in nvoc switch continuously between different levels when miner switch occurs with no kind of issues because not all P levels have customized OC values, Also, the endless loop in 3main can't lock gpu profile to P2, it has to be a compute job using the gpu to achive such a thing. That loop is there because nvOC is restarted otherwise.

**Comment by papampi** [comment]
Agree with all but now nvidia-smi shows persistence mode on and no warning that persistence mode is off and ....  when 3main starts. 
Dont see any reason not to enable it or at least add 1bash entry for those who want it.

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.77                 Driver Version: 390.77                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1070    On   | 00000000:01:00.0  On |                  N/A |
| 40%   65C    P2    99W / 100W |   3075MiB /  8119MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   1  P106-100            On   | 00000000:02:00.0 Off |                  N/A |
| 40%   64C    P0    75W /  75W |   2761MiB /  6080MiB |     98%      Default |
+-------------------------------+----------------------+----------------------+
|   2  P106-100            On   | 00000000:03:00.0 Off |                  N/A |
| 42%   70C    P0    73W /  75W |   2761MiB /  6080MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
```

**Comment by papampi** [comment]
I just added this line at the top of 3main:

`sudo nvidia-smi -pm 1`

**Comment by papampi** [comment]
And I feel first run after reboot is much faster, dont know ... maybe I'm wrong.

**Comment by LuKePicci** [comment]
The first 3main run after reboot is faster w.r.t next nvOC restarts because gpus are unused while nvOC is setting them up, same thing for tempcontrol. Obviously is damn faster w.r.t the old tempcontrol before this PR in all circumstances. It is ok to enable this only to remove warnings, but you should also take care of releasing when nvOC exits otherwise even restarting xorg does not allow drivers to be reloaded and it's not a good thing (again, persistence mode is designed to be used in HPC headless systems when there is no xorg or graphical jobs. but we have no nvOC shutdown handlers where to deal with this.

**Comment by papampi** [comment]
OK, lets sleep on it.
We will think about it for next release ;)

**Comment by LuKePicci** [comment]
TBH it is quite easy to do it when nvoc runs as a systemd, service, you can just enable/disable persistence in systemd start/stop recipes. And if the system supports systemd then the nvidia-persistenced is likely to be running as systemd service also, so you can easily enable this only if the persistece daemon is running. 
.

**Pull request merged**